### PR TITLE
Connection: disconnect user from wpcom before to do so locally

### LINKED ISSUE #19857: Connection: user token gets deleted locally before remotely when deleting user
#### Steps to reproduce the issue

1. Start with a site connected to WordPress.com, with 2 connected users: an admin (primary user for the connection) and an extra user.
2. At this point, when looking at your `user_tokens` Jetpack option, you'll see the 2 entries, one for each user id.
3. While logged in as the primary admin, go to Users in wp-admin
4. Delete the other user.

When you do so, this is triggered:
https://github.com/Automattic/jetpack/blob/3f71652367ec5cf0013e7d49f64320733578e291/projects/plugins/jetpack/class.jetpack.php#L679
https://github.com/Automattic/jetpack/blob/3f71652367ec5cf0013e7d49f64320733578e291/projects/plugins/jetpack/class.jetpack.php#L3507
https://github.com/Automattic/jetpack/blob/51edd8fd3b366e620c82624792fa52922513b23f/projects/packages/connection/src/class-manager.php#L767

At this point, we need to do 2 things:
- Disconnect the user locally
- Send a request to WordPress.com to disconnect the user there.

We start by disconnecting the user locally:
https://github.com/Automattic/jetpack/blob/51edd8fd3b366e620c82624792fa52922513b23f/projects/packages/connection/src/class-manager.php#L770
https://github.com/Automattic/jetpack/blob/9d0dac0ef082f21a19f36f24cbb82b12311bfc08/projects/packages/connection/src/class-tokens.php#L490-L492

At this point, the `user_tokens` Jetpack option does not include our deleted user info anymore.

Now, we go ahead and attempt to make a request to WordPress.com:

https://github.com/Automattic/jetpack/blob/51edd8fd3b366e620c82624792fa52922513b23f/projects/packages/connection/src/class-manager.php#L773-L774
When building the request, we look for the user token that will be used with the request:
https://github.com/Automattic/jetpack/blob/c52e77b1b4151e53a430b2d6ceb0adfb4e468faf/projects/packages/connection/src/class-client.php#L75-L78
https://github.com/Automattic/jetpack/blob/9d0dac0ef082f21a19f36f24cbb82b12311bfc08/projects/packages/connection/src/class-tokens.php#L359

But we don't have such a user token at this point anymore, so we bail:
https://github.com/Automattic/jetpack/blob/9d0dac0ef082f21a19f36f24cbb82b12311bfc08/projects/packages/connection/src/class-tokens.php#L371-L374

And thus the request to WordPress.com does not happen and the user does not get disconnected.

#### What I expected

I'm not sure what the solution would be here:

- Should we use the main user's token to make the request?
- Should we make the remote request first?

*******

Internal reference: 6613-gh-jpop-issues

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #19857

#### Changes proposed in this Pull Request:

* With this PR, we now attempt to make the request to disconnect the user from WordPress.com before we delete the connected data locally. This should fix the problem outlined in #19857.

#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

1. Start with a site connected to WordPress.com, with 2 connected users: an admin (primary user for the connection) and an extra user.
2. At this point, when looking at your `user_tokens` Jetpack option, you'll see the 2 entries, one for each user id.
3. While logged in as the primary admin, go to Users in wp-admin
4. Delete the other user.
5. Find the cache site in Network admin.
6. Ensure the wpcom user linked to the local user you just deleted got removed from Network admin.

### CODE REVIEW

**Review by kbrown9 — CHANGES_REQUESTED** [review]
I tested this using the provided testing instructions, and the user was disconnected as expected.

I think it would be nice to include a test that verifies that the local user token is not deleted when the remote disconnection fails. It would probably be an integration test, and the purpose of the test would be to catch the bug that is fixed by this PR. What do you think?

**Comment by jeherve** [comment]
> I think it would be nice to include a test that verifies that the local user token is not deleted when the remote disconnection fails. It would probably be an integration test, and the purpose of the test would be to catch the bug that is fixed by this PR. What do you think?

That makes sense. I'm not quite sure how to best go about this though. I pushed 39e791c0914e44213d28df911e04c8bd5037de88, but I don't think it's very useful. Right now the disconnection attempt fails in the test by default because the external XML-RPC request fails.

What should I do here? Should I aim to mock the response of the `unlink_user_from_wpcom` method just like I did in the other test, or somehow mock the response of the XML-RPC request itself, and add different use-cases to my test, with different outcomes based on whether the XML-RPC request is successful or not? I've tried to do the latter in 8faeaf386266d1f47cd523d40a0106664a119b96, but that doesn't work. Maybe because `Manager()` doesn't use my mocked `Jetpack_IXR_Client`? I'm not sure how to work around that, I assume we want to limit the mocks we have in our integration suite?

**Comment by kbrown9** [comment]
> What should I do here? Should I aim to mock the response of the `unlink_user_from_wpcom` method just like I did in the other test, or somehow mock the response of the XML-RPC request itself, and add different use-cases to my test, with different outcomes based on whether the XML-RPC request is successful or not? I've tried to do the latter in [8faeaf3](https://github.com/Automattic/jetpack/commit/8faeaf386266d1f47cd523d40a0106664a119b96), but that doesn't work. Maybe because `Manager()` doesn't use my mocked `Jetpack_IXR_Client`? I'm not sure how to work around that, I assume we want to limit the mocks we have in our integration suite?

Good questions! I pushed commit 2902741, which mocks the response of the XML-RPC request with a success or fail code. What do you think of that general approach? If you switch the order of the disconnects, so that the local disconnect happens before the remote disconnect, the test `Automattic\Jetpack\Connection\ManagerIntegrationTest::test_disconnect_user with data set "failure" (false, 2)` should fail.

**Comment by jeherve** [comment]
> What do you think of that general approach?

I like it, thanks for adding this all! This should now be ready for review then.

**Review by sdixon194 — APPROVED** [review]
this works for me! I'll hold off on merging until @kbrown9 can also take a look since the tests are added.

File: projects/packages/connection/tests/php/test_Manager_integration.php [hunk 1]
```diff
@@ -551,4 +551,146 @@ public function test_try_registration() {
...
+
+			$response['response']['code'] = 200;
+			return $response;
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Intercept the disconnect user API request sent to WP.com, and mock success response.
```
Comment (kbrown9) [inline projects/packages/connection/tests/php/test_Manager_integration.php:650]: I just noticed my own typo:
```suggestion
	 * Intercept the disconnect user API request sent to WP.com, and mock failure response.
```

**Review by sdixon194 — APPROVED** [review]

**Review by kbrown9 — APPROVED** [review]

**Pull request merged**